python3Packages.qiskit-aqua: init at 0.6.5

drewrisinger opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
Qiskit Aqua: An extensible library of quantum computing algorithms.

This commit follows the new Qiskit scheme of breaking one large package
into smaller packages (terra, aer, etc), and then having a single
meta-package "qiskit" that comprises them.

Broken out of #78772.
Waiting on #83306 #80662 #83372 to pass, builds cleanly when they're cherry-picked.
``python38Packages.qiskit-aqua`` will be broken due to upstream broken ``python38Packages.scikit-build`` (#83305 )

I disabled a lot of tests here, but there's still some 900 that run. So it's pretty well-tested.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

drewrisinger committed:
python3Packages.qiskit-aqua: init at 0.6.5

Qiskit Aqua: An extensible library of quantum computing algorithms.

This commit follows the new Qiskit scheme of breaking one large package
into smaller packages (terra, aer, etc), and then having a single
meta-package "qiskit" that comprises them.

drewrisinger commented:
@jonringer at this point it's just #83306 & #80662 left, which are in the same topic. I could wrap up the rest of the qiskit fixes in one PR, there's no new packages at this point. That way I could also demonstrate via bot that they all build. 

Summary of remaining qiskit fixes (all done, just didn't want to keep spamming PRs):
* Add ``pythonPackages.qiskit-[aer, ignis, aqua]``. See #83306, #80662 
* Delete ``pythonPackages.IBMQuantumExperience`` (superseded by ``pythonPackages.qiskit-ibmq-provider`` in merged #83320 )
* bump ``pythonPackages.qiskit`` -> 0.16.2, remove broken tag, make it a superset of the ``qiskit-[terra,aer,ignis,aqua]`` packages. #78772

drewrisinger commented:
@GrahamcOfBorg eval
running build locally (both via ``nixpkgs-review`` and ``nix-build``). Will update with status

drewrisinger commented:
Update: builds clean via ``nixpkgs-review pr 83447``.

2 package built:
python37Packages.qiskit-aqua python38Packages.qiskit-aqua

bcdarwin reviewed (commented):
commit/PR should refer to `python3Packages`

bcdarwin commented on pkgs/development/python-modules/qiskit-aqua/default.nix:
```diff
@@ -0,0 +1,104 @@
+    "--ignore-glob=*qgan.py"
+  ];
+  disabledTests = [
+    # Disabled due to missing pyscf
```
If `pyscf` can't be added to `checkInputs` for some reason, probably worth noting.

drewrisinger replied:
I tried packaging pyscf in #78872, and myself & jonringer ended up deciding that it's being too difficult/ugly to package (issue mostly coming down to ``pyscf`` tests being very slow, failures difficult to diagnose, and long list of disabled tests that is hard to maintain). I can add note to that effect in code.

bcdarwin commented:
Looks fine to me, will try to build later.

timokau commented on pkgs/development/python-modules/qiskit-aqua/default.nix:
```diff
@@ -0,0 +1,104 @@
+  postPatch = ''
+    # Make pyscf optional, can be included at runtime via pip shell
+    substituteInPlace setup.py \
+      --replace "pyscf; sys_platform == 'linux' or (python_version < '3.8' and sys_platform != 'win32')" ""
```
How will that work at runtime? Imagine you're a user that just installed this package from nixpkgs and expects it to work. Will there be unexpected crashes? If so, we should patch in a proper error message explaining the situation.

drewrisinger replied:
Good point. I'd be fine adding error message on import. Will try to patch in.

timokau commented on pkgs/development/python-modules/qiskit-aqua/default.nix:
```diff
@@ -0,0 +1,104 @@
+  # Pypi's tarball doesn't contain tests
+  src = fetchFromGitHub {
+    owner = "Qiskit";
+    repo = pname;
```
```suggestion
    repo = "qiskit-aqua";
```

This is an "accidental similarity" (my own term, there's probably a real name for this but I haven't found it yet). While `repo` and `pname` are the same here, they don't necessarily change together. So it would be better to just repeat the name here.

drewrisinger replied:
ack

drewrisinger replied:
@timokau re "accidental similarity": I suggest "coincidental similarity"? Or "naming hash collision" :D

drewrisinger commented:
Let's see if we can actually build it now via borg after #84124...
@GrahamcOfBorg build python37Packages.qiskit-aqua
@GrahamcOfBorg build python38Packages.qiskit-aqua
Took everyone's code review suggestions into account, and added an ugly-ish description of why pyscf #78872 isn't included, along with an ImportWarning when someone would try to use the appropriate chemistry section of this package.

timokau commented on pkgs/development/python-modules/qiskit-aqua/default.nix:
```diff
@@ -0,0 +1,116 @@
+    # Add ImportWarning when running qiskit.chemistry (pyscf is a chemistry package) that pyscf is not included
+    echo -e "\nimport warnings\ntry: import pyscf;\nexcept:\n    " \
+      "warnings.warn('pyscf is not supported on Nixpkgs so some qiskit features will fail." \
+        "You must install it yourself via pip or add it to your environment from the Nix User Repository.', ImportWarning)\n" \
```
It would be nice to give the user some starting point in case they want to find out more. You could link either this PR or the one where you tried to package `pyscf`.

drewrisinger replied:
Ack. Mentioned both in comments, this in the user-facing text.

timokau reviewed (changes requested):
Looks like the aarch64 build is failing. Please restrict [`meta.platforms`](https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes) to x86-linux.

drewrisinger committed:
python3Packages.qiskit-aer: only enable on x86_64

Build was failing on ofborg on platforms.aarch64 due to
missing muparserx library built for aarch64. Added notes about
this issue & when build could be expanded.

drewrisinger commented:
@GrahamcOfBorg build python37Packages.qiskit-aqua python38Packages.qiskit-aqua
Left ``python3Packages.pyscf`` discussions "unresolved" above, even though they're actually resolved, as future notes for anyone who finds this issue linked in description.

Traced failing build to ``pythonPackages.qiskit-aer`` which had original unfree dependency and so couldn't test against ``aarch64`` on ofborg. Left notes about what the issue is; fully resolving it would require packaging https://github.com/beltoforion/muparserx, which I'll leave for another day.

timokau commented:
Looks good, thanks!
